New addon - custom block shape

lisa-wolfgang opened the pull request:
![image](https://user-images.githubusercontent.com/43426138/111520301-f80be180-8725-11eb-880b-d041217d6034.png)
_A screenshot of the latest version with the 2.0 Blocks preset_

**Resolves**

<!-- Which issue(s) does this pull request fix or resolve? -->

Resolves #592 (and its 2 duplicates)
Resolves #953

**Changes**

<!-- Please describe the changes you've made. -->

This is a modification of @SheepTester's "Thinner Scratch blocks" -- the block padding, corner radius, and notch height can be changed by the user via settings in the corresponding addon, from 50% to 200% for padding, 0% to 300% for radius, and 0% to 150% for notch. The default 100% setting keeps the size the same as usual. Presets are also provided for commonly desired themes like Scratch 2.0.

**Reason for changes**

<!-- Why should these changes be made? -->

Many people would like a way to decrease block size for aesthetic reasons and because it allows more blocks to be seen on the screen at once without decreasing text size. The other options also cater to these people.

**Known issues**

<!-- Have you tested this pull request? If so, how? -->
- Notch sizes greater than 100% have a very small line protruding into the notch (barely noticeable)
- Changing the padding size doesn't change the vertical positions of blocks (a warning is given about this)

There are probably some other issues since this is my first addon contribution and I don't know the standards very well.

Endermaster3690 commented:
I support this, I've seen topics on Scratch talking about blocks being to big, so we needed a way to change block size. This is really cool, keep it worked on if possible!

lisa-wolfgang commented:
Just fixed some major issues.
Before:
![image](https://user-images.githubusercontent.com/43426138/110578499-98f90c00-812a-11eb-9647-ecc0bee66ed0.png)
After:
![image](https://user-images.githubusercontent.com/43426138/110694047-c2a74700-81ad-11eb-8b1b-ca4a7ae47ebe.png)
- Fixed notch misalignment
- Fixed text box text not being centered
- Fixed gap between loop/statement edge and its contents

TheColaber commented:
Please enable prittier at https://github.com/lisa-wolfgang/ScratchAddons/actions

Endermaster3690 commented:
> Just fixed some major issues.

> * Fixed notch misalignment
> * Fixed text box text not being centered
> * Fixed gap between loop/statement edge and its contents

Great! Keep working, but another major bug is the Boolean still doing that.

Endermaster3690 commented:
> Another update:
> 
> * Added corner radius option (0% to 300%)
> * Fixed boolean sizing issue
> 
> Since I don't think there's anything left that I can fix, I'm opening it up for review.

I can't wait! Come on world languages, you gotta accept this!

TheColaber commented:
Sometimes the flyout blocks are bigger than the blocks that get dragged out... this seems to be the case with all blocks that have inputs...
![image](https://user-images.githubusercontent.com/72760579/110721639-e9ca3c80-81de-11eb-8014-b6cb23081040.png)
Reduce spacing:
![image](https://user-images.githubusercontent.com/72760579/110721679-fbabdf80-81de-11eb-83f3-3046a63be254.png)
Hmm...
![image](https://user-images.githubusercontent.com/72760579/110721716-0bc3bf00-81df-11eb-9c4f-f3c38d984877.png)
Conflicts with cat-blocks right now, but this is my problem, don't worry, see #1809
![image](https://user-images.githubusercontent.com/72760579/110721880-5a715900-81df-11eb-9b1c-b3585df8ee08.png)

Endermaster3690 commented:
> Hmm...
> ![image](https://user-images.githubusercontent.com/72760579/110721716-0bc3bf00-81df-11eb-9c4f-f3c38d984877.png)

You make a chungus big block and see the arrow at the top of the input XD Gotta get fixed fast.

Hans5958 commented:
> Please enable prettier at https://github.com/lisa-wolfgang/ScratchAddons/actions

@TheColaber Reminder: You actually don't need to activate the action. As long as the check passes (if you use Prettier on your enviroment), it's good. I think it is quite misleading to tell people to turn it on, instead of fixing the checks.

lisa-wolfgang commented:
New stuff -- notch height is now an option!
I'll also try to fix some more bugs if I can.

lisa-wolfgang commented:
Finally fixed that pesky dropdown bug! (big thanks to GarboMuffin for that :D)
In addition, padding size can now be increased to 200%.

lisa-wolfgang commented:
See #1836

WorldLanguages commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,233 @@
+  var BlocklyInstance = await addon.tab.traps.getBlockly();
+
+  (function (Blockly) {
+    "use strict";
```
Not needed, already on strict mode

WorldLanguages commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,57 @@
+      "max": 200
+    }
+  ],
+  "traps": true,
```
Not needed

WorldLanguages commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,57 @@
+      "type": "integer",
+      "min": 0,
+      "default": 100,
+      "max": 200
```
Is max 150 or 200?

WorldLanguages commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,57 @@
+  "info": [
+    {
+      "type": "notice",
+      "text": "Setting the padding size to a smaller value may decrease usability.",
```
Can you elaborate what "usability" means on here?

lisa-wolfgang replied:
I'm actually just going to remove that message and replace it with a warning about block positioning.

WorldLanguages commented:
1. Would be great to have the default values be close to the looks of Scratch 2.0 blocks
2. Stuff is weird in the scripts tab after changing a setting (e.g. padding), until you force that tab to load again by going into costumes/sounds and then back. Can this be automated? Either by emulating clicking into one of these tabs, or some call to `Blockly.*` that forces a rerender (preferred)

Endermaster3690 commented:
> Would be great to have the default values be close to the looks of Scratch 2.0 blocks

How about like with the block color changing, we have a templates thing that has 2.0 colors and have a reset button next to all the values?

lisa-wolfgang commented:
> > Would be great to have the default values be close to the looks of Scratch 2.0 blocks
> 
> How about like with the block color changing, we have a templates thing that has 2.0 colors and have a reset button next to all the values that set them to the 3.0 values?

@Endermaster3690 Actually, I just added that in the most recent update!

Endermaster3690 commented:
> > > Would be great to have the default values be close to the looks of Scratch 2.0 blocks
> > 
> > 
> > How about like with the block color changing, we have a templates thing that has 2.0 colors and have a reset button next to all the values that set them to the 3.0 values?
> 
> @Endermaster3690 Actually, I just added that in the most recent commit!

Great, also, you made a commit about fixing define hat corners.
How about a check box that makes define hats look like 2.0?

GarboMuffin commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,108 @@
+  "credits": [
+    {
+      "name": "SheepTester",
+      "link": "https://sheeptester.github.io/hello-world/userscripts/"
```
might be more meaningful to point to https://scratch.mit.edu/users/Sheep_maker/, at least I'll probably change it to that for TW because I don't want links directly to userscripts because of Scratch policy

lisa-wolfgang replied:
I'll change that -- other addons are linking directly to the userscript, so that's why I chose to include that.

GarboMuffin commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,247 @@
+  var BlocklyInstance = await addon.tab.traps.getBlockly();
+
+  (function (Blockly) {
+    const workspace = Blockly.getMainWorkspace();
```
you should be using Blockly.getMainWorkspace() every time you need the workspace as it will switch whenever the user enters or leaves the editor

lisa-wolfgang replied:
I'll change it, but don't JS constants store pointers to the value rather than actually freezing the value?

RedGuy12 replied:
its confusing

it stores a pointer only for objects (including arrays, dom, classes, etc). Other types it stores the actual value.

But I think since `Blockly.getMainWorkspace()` is a function, the value is not stored anywhere yet so it has to make a new pointer.

I think

lisa-wolfgang commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,251 @@
+
+  (function (Blockly) {
+    const workspace = Blockly.getMainWorkspace();
+    const BlockSvg = Blockly.getMainWorkspace().newBlock().constructor;
```
Blame SheepTester for that lol

GarboMuffin commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,246 @@
+    var cornerSize = 1;
+    var notchSize = 1;
+
+    function updateAllBlocks() {
+      if (vm.editingTarget) {
+        vm.emitWorkspaceUpdate();
+        Blockly.getMainWorkspace().getToolbox().flyout_.workspace_.clear();
+        Blockly.getMainWorkspace().getToolbox().refreshSelection();
+      }
+    }
```
So apparently this still causes a bunch issues of varying severity, my bad.

I think they're all fixed here, for real this time:

```suggestion
    function updateAllBlocks() {
      const workspace = Blockly.getMainWorkspace();
      if (workspace) {
        if (vm.editingTarget) {
          vm.emitWorkspaceUpdate();
        }
        const flyout = workspace.getFlyout();
        if (flyout) {
          const flyoutWorkspace = flyout.getWorkspace();
          Blockly.Xml.clearWorkspaceAndLoadFromXml(Blockly.Xml.workspaceToDom(flyoutWorkspace), flyoutWorkspace);
          workspace.getToolbox().refreshSelection();
          workspace.toolboxRefreshEnabled_ = true;
        }
      }
    }
```

RedGuy12 commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,112 @@
+      "max": 300
+    },
+    {
+      "name": "Notch height (0-150%)",
```
The text in parentheses isn't needed I don't think since the element has min and max attrs. But if you decide to keep it, I don't think there should be a percent there

lisa-wolfgang replied:
I think it's helpful to know the possible parameters that can be entered, and the percent helps people visualize the effect that the setting will have.

TheColaber replied:
I agree with @lisa-wolfgang... users aren't expected to inspect the page to figure that out.

RedGuy12 commented on addons/custom-block-shape/addon.json:
```diff
@@ -0,0 +1,112 @@
+      }
+    },
+    {
+      "name": "Bigger",
```
I don't think many people would like this, there's already a lot of criticism about the large blocks as it is

lisa-wolfgang replied:
OK.

RedGuy12 commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,246 @@
+export default async function ({ addon, global, console }) {
+  var multiplier = addon.settings.get("paddingSize") / 100;
```
Its overridden on L22?

lisa-wolfgang replied:
That's a leftover from some attempt at something a while ago, so it can go back down by the other setting variables.

RedGuy12 commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,246 @@
+
+    const { GRID_UNIT } = BlockSvg;
+    var cornerSize = 1;
+    var notchSize = 1;
```
ditto

lisa-wolfgang replied:
The 1s are also leftovers and can be removed.

RedGuy12 commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,246 @@
+      BlockSvg.NOTCH_HEIGHT = 2 * GRID_UNIT * multiplier * notchSize;
+      BlockSvg.NOTCH_START_PADDING = 3 * GRID_UNIT; //* multiplier
+      BlockSvg.ICON_SEPARATOR_HEIGHT = 10 * GRID_UNIT * multiplier;
+      BlockSvg.NOTCH_PATH_LEFT =
```
Maybe use a template string?

lisa-wolfgang replied:
?

SheepTester replied:
They mean something like

```suggestion
      BlockSvg.NOTCH_PATH_LEFT = `c 2,0 3, ${1 * notchSize} 4, ${2 * notchSize} etc`
```

lisa-wolfgang replied:
Ah, I see. It was already in the + format, so I figured it would be easier to just leave it like that.

RedGuy12 commented on addons/custom-block-shape/userscript.js:
```diff
@@ -0,0 +1,246 @@
+
+    addon.settings.addEventListener("change", function () {
+      applyChanges();
+      updateAllBlocks();
```
Use `applyAndUpdate()`

superscratch444tre commented:
Look:
![image](https://user-images.githubusercontent.com/69812464/111456292-ba4f8e80-870e-11eb-9566-ef051b70a028.png)

There's a tiny dot....
padding 100, corner 54, notch 0

lisa-wolfgang commented:
Actually, thanks to GarboMuffin (again) I found the property that needs to be changed. Pushing soon!

Endermaster3690 commented:
Is there notch width?

WorldLanguages commented:
2.0 is that you? 👀 
![image](https://user-images.githubusercontent.com/17484114/111691698-55c72900-880d-11eb-84c2-7ec095466af0.png)

TheColaber commented:
The future... is the past.

Chikorita-Lover commented:
Hooray, it has been merged! 🎉

lisa-wolfgang commented:
> Is there notch width?

Not right now, but perhaps in the future!

Endermaster3690 commented:
> > Is there notch width?
> 
> Not right now, but perhaps in the future!

cant wait for this to come in 1.11.1!

superscratch444tre commented:
> 2.0 is that you? 👀
> ![image](https://user-images.githubusercontent.com/17484114/111691698-55c72900-880d-11eb-84c2-7ec095466af0.png)

Combine this with https://mxmou.github.io/customize-scratch/scratch2Theme.user.js and it will really be scratch 2.0

Endermaster3690 commented:
> > > Is there notch width?
> > 
> > 
> > Not right now, but perhaps in the future!
> 
> cant wait for this to come in 1.11.1!

my micro brain just realized it's 1.11.2 and its not here
hm guess its coming 1.12
